Modularize imageTask and lock the mutex during the entire image acquisition.

gustavosr8 opened the pull request:
These changes were made to reduce the dead time seen in the operation of the detector after the acquisition of an image. We found that the competition between the acquisition task and the temperature reading task was causing some unnecessary delay in image acquisition, so we decided that the acquisition task should hold the mutex throughout the acquisition operation and release it at the end.

gustavosr8 committed:
Remove commented code.

There are several parts of this code that are commented and degrade
readability. Removing it isn't harmful, since it will always be
accessible in the git history.

gustavosr8 committed:
Move capture parameters configuration to an auxiliary function.

The capture parameters configuration is embedded directly in the
imageTask logic, reducing readability and reusability. The error
handling overwrites the result variable multiple times without checking
its state, which can lead to lost error information.

Extract the capture parameters configuration into a dedicated auxiliary
function to improve modularity and clarify the code structure. Modify
error handling to accumulate errors instead of overwriting the result
variable, ensuring all errors are accounted for and reported correctly.

gustavosr8 committed:
Split the imageTask into capture and non-capture routine.

The imageTask function handles both capture and non-capture operations
in a single routine, making the logic complex and harder to maintain.
The operations for configuring camera parameters, checking for
acquisition requests, and handling simulations are mixed together,
reducing clarity.

Split imageTask into separate capture and non-capture routines. The
non-capture routine configures the camera parameters and checks for
acquisition requests. The capture routine handles the same configuration
(keeping consistent with the previous behavior), checks for simulations,
performs the actual capture, and verifies whether to stop or continue
acquisition at the end. This separation improves readability and
simplifies the task flow.

gustavosr8 committed:
Unify stopEvent handling in acquisition loop.

Simplify stopEvent handling by unifying epicsEventTryWait and
epicsEventWaitWithTimeout calls, as both are equivalent when delay == 0
[1]. Now, if stop is received before image processing, the image is
processed before changing ADStatus. Also, the port will always be
unlocked and locked, which is not a problem since the write functions are
quite short.

Since delay enforces the acquisition period, a negative delay only
happens if the acquisition is late, so all waits should be non-blocking
(delay=0) to avoid delaying it further.

Removing the second block which checks the stopEvent doesn't affect
behavior, as acquire isn't checked again before being reset.

[1] https://docs.epics-controls.org/projects/base/en/latest/epicsEvent_h.html#_CPPv425epicsEventWaitWithTimeout12epicsEventIdd

gustavosr8 committed:
Turn the acquisition process into a loop and unify end points.

The acquisition process handles endpoint conditions inconsistently,
making the code harder to follow and maintain. Refactor the acquisition
process into a loop with a unified endpoint check to simplify control
flow and make the end of acquisition easier to handle.

Due to an early return, detector state goes to idle rather than aborted
when the ADImageSingle is in use. To keep this behavior without this
early return, check if the imageMode is ADImageMultiple (that is, not
ADImageSingle nor ADImageContinuous) to change to aborted; otherwise, go
to idle.

gustavosr8 committed:
Turn the non acquisition process into a loop.

Refactor the process into a loop to improve clarity and simplify
handling the start of acquisition.

Note that now it doesn't check for !_exiting anymore, what shouldn't be
a problem since Shutdown() won't wait for any threads at all, and the
thread itself doesn't seem to require terminating in any specific state.

gustavosr8 committed:
Extract capture process into a method.

The capture process is currently embedded within the main logic, making
it complex and harder to maintain. The sequence of operations, including
setting camera status, retrieving exposure parameters, capturing the
image, and handling various image modes, is not well encapsulated.

A new function, doAcquisition, was introduced to manage the full
acquisition flow. It handles camera state transitions such as
ADStatusAcquire, ADStatusReadout, and ADStatusWaiting, opens and closes
the shutter at appropriate stages, and supports different image modes
like ADImageSingle, ADImageMultiple, and ADImageContinuous. Even though
imageTask still sets ADAcquire itself, controlling ADStatus happens in
doAcquisition, based on the value of acquisition_done that's being set;
this is necessary because the reason for ending the acquisition isn't
available outside of doAcquisition. It also manages simulation behavior
by sleeping for the acquisition time and monitoring for interrupts.

The mutex will be acquired before calling configureCaptureParams() and
released after it returns. The GetImage() function ensures that the
mutex is locked during image retrieval.

The simulation mode sleeps for the acquireTime, to simulate being busy,
and watches for a stopEvent, setting the status to ADStatusAborted if
stopEvent has been received.

gustavosr8 committed:
Extract pre capture process into a method.

Create a dedicated function to handle the pre-capture process. The
function will acquire the mutex before calling configureCaptureParams()
and release it afterward.

gustavosr8 committed:
Lock the mutex during the entire acquisition.

In this way, performance isn't affected by concurrent tasks - i.e.
temperature reading.

gustavosr8 committed:
Update ADAcquire to keep consistency with the control variable.

The ADAcquire parameter was only configured to false once, at the end of
acquisition, and then never back to true. Use setIntegerParam to update
its value whenever the control variable is updated.

Note that immediate callbacks are missing now and will be added in the
subsequent commits.

ericonr commented:
@stuwilkins and @MarkRivers , could you take a look at this?

gustavosr8 committed:
Document best use of mutex to prevent dead locks.

gustavosr8 committed:
Use locks in temperature callbacks to make sure it works properly.

henriquesimoes commented on rixscamApp/src/rixscam.cpp:
```diff
@@ -1263,9 +1263,12 @@ void xcamCamera::temperatureTask(void)
 		setDoubleParam(ADTemperatureActual, actualTempCelcius);
 
 		callParamCallbacks();
+		this-unlock();
```
Broken deference to unlock: `s/this-unlock/this->unlock`.

gustavosr8 replied:
ops, missed it. Thanks!

henriquesimoes reviewed (commented):
I've found a typo in your latest commit.

gustavosr8 commented:
Code here was previously reviewed by @ericonr and @henriquesimoes in a private gitlab repo and has been used by the IPE beamline on Sirius since January/2025 without any problems.

MarkRivers reviewed (commented):
In line 887 you are releasing the lock before calling doCallbacksGenericPointer.  Most drivers used to do that, but it was based on a faulty diagnosis of an issue.  It can lead to issues, because other threads can do things when the lock is released.  You should see if this is really needed because of something unique to this driver.

MarkRivers commented:
This is difficult to review because there were major changes to the whitespace in addition to code changes.  Is there a way to have Github do the equivalent of "git diff -w -b"?

MarkRivers commented:
These are comments from RELEASE.md files for other drivers showing that releasing the lock around calls to doCallbacksGenericPointer was removed:
```
./ADDexela/RELEASE.md:* Removed calls to unlock()/lock() around doCallbacksGenericPointer, not needed.
./ADPointGrey/RELEASE.md:* Removed code that released the lock around the calls to doCallbacksGenericPointer.
./ADPerkinElmer/RELEASE.md:* Removed calls to release lock around calls to doCallbacksGenericPointer.
./ADmarCCD/RELEASE.md:  gating, high-speed series acquisition with either external trigger or internal clock, and support
./ADURL/RELEASE.md:* Remove unlock() and lock() around call to doCallbacksGenericPointer. It is not needed and can cause problems.
./pvaDriver/RELEASE.md:* Removed unlocking around doCallbacksGenericPointer, not needed and can be unsafe.
./ADCore/RELEASE.md:* Previously all plugins were releasing the asynPortDriver lock when calling doCallbacksGenericPointer().
./ADCore/RELEASE.md:  This was based on a very old observation of a deadlock problem if the the lock was not released.  Releasing
./ADCore/RELEASE.md:  the lock causes serious problems with plugins running multiple threads, and probably was never needed.  Most
./ADCore/RELEASE.md:  The lock is no longer released when calling doCallbacksGenericPointer().  The simDetector driver has also been modified
./ADCore/RELEASE.md:  to no longer release the lock when calling plugins with doCallbacksGenericPointer(), and all other drivers should be
./ADProsilica/RELEASE.md:* Removed calls to unlock()/lock() around doCallbacksGenericPointer, not needed.
```
Other drivers also had this change made, but it was not discussed in the RELEASE.md file.

henriquesimoes commented:
> Is there a way to have Github do the equivalent of "git diff -w -b"?

The best it can do is to ["Hide whitespace"](https://github.blog/changelog/2021-10-14-hiding-whitespace-is-now-remembered-for-each-pull-request/). Not sure if that fulfills `-w`, `-b`, or both.

For this patch set, I highly recommend reviewing it commit-wise. I've done that locally instead of the forge's UI to use `git diff --color-moved --color-moved-ws=ignore-space-change`. This helped me a lot to verify what gets moved where.

gustavosr8 commented:
> In line 887 you are releasing the lock before calling doCallbacksGenericPointer.

Although it appears as a new commit, it was already in the master version of the code (https://github.com/areaDetector/ADRIXSCam/blob/master/rixscamApp/src/rixscam.cpp#L1264). Since the comment mentions the deadlock problem, and since, as you mentioned, we now know it was an incorrect diagnosis, I can add a new commit with this change.

gustavosr8 committed:
Removed calls to unlock()/lock() around doCallbacksGenericPointer.

As mentioned in the ADCore release notes [1], release the lock causes
problems and is no longer recomended.

[1] https://github.com/areaDetector/ADCore/blob/3acafd04140730fc7a0108178a3e2597c57c4901/RELEASE.md?plain=1#L1015

MarkRivers commented:
> The best it can do is to ["Hide whitespace"](https://github.blog/changelog/2021-10-14-hiding-whitespace-is-now-remembered-for-each-pull-request/). Not sure if that fulfills -w, -b, or both.

Thanks, I had missed that "Hide whitespace" option.

> For this patch set, I highly recommend reviewing it commit-wise. 

Yes, that is much more useful.

I am happy with this PR once the unlock around doCallbacksGenericPointer is fixed.
